# Exploring Hacker News Posts

## Introduction

In this project, we'll compare two different types of posts from Hacker News, a popular site where technology related stories (or 'posts') are voted and commented upon. The two types of posts we'll explore begin with either Ask HN or Show HN.

Through this project we will try to determine the following:
1. Do Ask HN or Show HN receive more comments on average?
2. Do posts created at a certain time receive more comments on    average?

The dataset we are using is available here: https://www.kaggle.com/datasets/hacker-news/hacker-news-posts

## Reading the Data into Python

First, we'll read in the data and remove the headers.

In [64]:
#Reading the data and converting iat to a list of lists
from csv import reader
open_file = open("hacker_news.csv")
read_file = reader(open_file)
hn = list(read_file)
print(hn[:5])
    
    

[['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at'], ['12224879', 'Interactive Dynamic Video', 'http://www.interactivedynamicvideo.com/', '386', '52', 'ne0phyte', '8/4/2016 11:52'], ['10975351', 'How to Use Open Source and Shut the Fuck Up at the Same Time', 'http://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-the-same-time/', '39', '10', 'josep2', '1/26/2016 19:30'], ['11964716', "Florida DJs May Face Felony for April Fools' Water Joke", 'http://www.thewire.com/entertainment/2013/04/florida-djs-april-fools-water-joke/63798/', '2', '1', 'vezycash', '6/23/2016 22:20'], ['11919867', 'Technology ventures: From Idea to Enterprise', 'https://www.amazon.com/Technology-Ventures-Enterprise-Thomas-Byers/dp/0073523429', '3', '1', 'hswarna', '6/17/2016 0:01']]


In [65]:
#Removing Headers and assigning to a different list
headers = hn[0]
hn= hn[1:]
print(hn[:5])


[['12224879', 'Interactive Dynamic Video', 'http://www.interactivedynamicvideo.com/', '386', '52', 'ne0phyte', '8/4/2016 11:52'], ['10975351', 'How to Use Open Source and Shut the Fuck Up at the Same Time', 'http://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-the-same-time/', '39', '10', 'josep2', '1/26/2016 19:30'], ['11964716', "Florida DJs May Face Felony for April Fools' Water Joke", 'http://www.thewire.com/entertainment/2013/04/florida-djs-april-fools-water-joke/63798/', '2', '1', 'vezycash', '6/23/2016 22:20'], ['11919867', 'Technology ventures: From Idea to Enterprise', 'https://www.amazon.com/Technology-Ventures-Enterprise-Thomas-Byers/dp/0073523429', '3', '1', 'hswarna', '6/17/2016 0:01'], ['10301696', 'Note by Note: The Making of Steinway L1037 (2007)', 'http://www.nytimes.com/2007/11/07/movies/07stein.html?_r=0', '8', '2', 'walterbell', '9/30/2015 4:12']]


## Extracting Ask HN and Show HN Posts

As we are only interested in Ask HN and Show HN posts,let us extract them into separate list of lists for analysis

In [66]:
#Refining the data to extract Ask HN and Show HN posts
ask_posts = []
show_posts = []
other_posts = []

In [67]:
for row in hn:
    title = row[1]
    if title.lower().startswith('ask hn'):
        ask_posts.append(row)
    elif title.lower().startswith('show hn'):
        show_posts.append(row)
    else:
        other_posts.append(row)


In [68]:
print(len(ask_posts))
print(len(show_posts))
print(len(other_posts))

1744
1162
17194


## Calculating the Average Number of Comments for Ask HN and Show HN Posts

Now that we separated ask posts and show posts into different lists, we'll calculate the average number of comments each type of post receives.

In [69]:
#Calculaing average number of comments in ask posts
total_ask_comments = 0
i = 0
for post in ask_posts:
    total_ask_comments += int(post[4])
    i += 1
avg_ask_comments = total_ask_comments/i
print(avg_ask_comments)

14.038417431192661


In [70]:
#Calculaing average number of comments in show posts
total_show_comments = 0
j = 0 
for post in show_posts:
    total_show_comments += int(post[4])
    j += 1
avg_show_comments = total_show_comments/j
print(avg_show_comments)

10.31669535283993


From the analysis done above, it can be inferred that on an average, ask posts have more comments compared to show posts.
While show posts have approximately 10 comments per post, ask posts have around 14 commetns per post

Based on this analysis, we will focus our analysis on ask posts as they are likely to recieve more comments

## Finding the Amount of Ask Posts and Comments by Hour Created

Now, we shall determine if we can maximize the amount of comments an ask post receives by creating it at a certain time.We will find the amount of ask posts created during each hour of day, along with the number of comments those posts received. Then, we'll calculate the average amount of comments ask posts created at each hour of the day receive.

In [71]:
#Calculating the number of ask posts created by hour and the number of comments recieved
import datetime as dt
result_list = []
for row in ask_posts:
    create_time = row[6]
    comment_no = int(row[4])
    result_list.append([create_time,comment_no])

counts_by_hour = {}
comments_by_hour = {}
for item in result_list:
    time = item[0]
    comment = item[1]
    time = dt.datetime.strptime(time,'%m/%d/%Y %H:%M').strftime('%H')
    if time in counts_by_hour:
        counts_by_hour[time] += 1
        comments_by_hour[time] += comment
    else:
        counts_by_hour[time] = 1
        comments_by_hour[time] =comment   
    

In [72]:
#Calculating average comments by hour
avg_by_hour = []
for hr in comments_by_hour:
    avg_by_hour.append([hr,(comments_by_hour[hr]/counts_by_hour[hr])])
print(avg_by_hour)

[['09', 5.5777777777777775], ['13', 14.741176470588234], ['10', 13.440677966101696], ['14', 13.233644859813085], ['16', 16.796296296296298], ['23', 7.985294117647059], ['12', 9.41095890410959], ['17', 11.46], ['15', 38.5948275862069], ['21', 16.009174311926607], ['20', 21.525], ['02', 23.810344827586206], ['18', 13.20183486238532], ['03', 7.796296296296297], ['05', 10.08695652173913], ['19', 10.8], ['01', 11.383333333333333], ['22', 6.746478873239437], ['08', 10.25], ['04', 7.170212765957447], ['00', 8.127272727272727], ['06', 9.022727272727273], ['07', 7.852941176470588], ['11', 11.051724137931034]]


In [73]:
#Sorting and printing the values
swap_avg_by_hour = []
for row in avg_by_hour:
    swap_avg_by_hour.append([row[1],row[0]])
print (swap_avg_by_hour)

[[5.5777777777777775, '09'], [14.741176470588234, '13'], [13.440677966101696, '10'], [13.233644859813085, '14'], [16.796296296296298, '16'], [7.985294117647059, '23'], [9.41095890410959, '12'], [11.46, '17'], [38.5948275862069, '15'], [16.009174311926607, '21'], [21.525, '20'], [23.810344827586206, '02'], [13.20183486238532, '18'], [7.796296296296297, '03'], [10.08695652173913, '05'], [10.8, '19'], [11.383333333333333, '01'], [6.746478873239437, '22'], [10.25, '08'], [7.170212765957447, '04'], [8.127272727272727, '00'], [9.022727272727273, '06'], [7.852941176470588, '07'], [11.051724137931034, '11']]


We shall now sort the the values for the number of comments in descending order and print them as a list of lists. This will help us determine the optimum time period for posting ask posts.

In [74]:
sorted_swap = sorted(swap_avg_by_hour,reverse = True)
print(sorted_swap[:5])

[[38.5948275862069, '15'], [23.810344827586206, '02'], [21.525, '20'], [16.796296296296298, '16'], [16.009174311926607, '21']]


In [85]:
print ("Top 5 Hours for Ask Posts Comments")
text = "{hour}: {comment:,.2f} average comments per post"
for avg,hr in sorted_swap[:5]:
    print(text.format(hour = dt.datetime.strptime(hr, "%H").strftime("%H:%M"),comment= avg))

Top 5 Hours for Ask Posts Comments
15:00: 38.59 average comments per post
02:00: 23.81 average comments per post
20:00: 21.52 average comments per post
16:00: 16.80 average comments per post
21:00: 16.01 average comments per post


From the data above,the hour that receives the most comments per post on average is 15:00, with an average of 38.59 comments per post. There's about a 60% increase in the number of comments between the hours with the highest and second highest average number of comments.

According to the data set documentation, the timezone used is Eastern Time in the US. So, we could also write 15:00 as 3:00 pm est.

## Conclusion

In this project, we analyzed ask posts and show posts to determine the  type of post that receive the most comments on average. we also determined the time that receives the most comments per post. Based on our analysis, to maximize the number of comments a post receives, we'd recommend the post be categorized as ask post and created between 15:00 and 16:00 (3:00 pm est - 4:00 pm EST).

However, it should be noted that the data set we analyzed excluded posts without any comments.
Keeping that in mind, it's more accurate to say that of the posts that received comments, ask posts received more comments on average and amoong those, ones created between 15:00 and 16:00 (3:00 pm est - 4:00 pm est) received the most comments on average.